File: 307. Range Sum Query - Mutable/307. Range Sum Query - Mutable.py
class SegmentTree:

    def __init__(self,L,R,total):
        self.total = total
        self.L=L
        self.R=R
        self.left=None
        self.right=None

    @staticmethod
    def build(nums,L,R):
        if L==R:
            return SegmentTree(L,R,nums[R])
        
        M = (L+R)//2
        root = SegmentTree(L,R,0)
        root.left = SegmentTree.build(nums,L,M)
        root.right = SegmentTree.build(nums,M+1,R)
        root.total = root.left.total +root.right.total
        return root
    
    def update(self,index,val):
        if self.L == self.R:
            self.total = val
            return 
        M = (self.L + self.R)//2
        if index<=M:
            self.left.update(index,val)
        else:
            self.right.update(index,val)
        self.total = self.left.total +self.right.total

    def queryRange(self,L,R):
        if L== self.L and self.R==R:
            return self.total
        M = (self.L + self.R) // 2
        if L > M:
            return self.right.queryRange(L, R)
        elif R <= M:
            return self.left.queryRange(L, R)
        else:
            return (self.left.queryRange(L, M) +
                    self.right.queryRange(M + 1, R))



class NumArray:

    def __init__(self, nums: List[int]):
      self.segement = SegmentTree.build(nums,0,len(nums)-1)
        

    def update(self, index: int, val: int) -> None:
        self.segement.update(index,val)

    def sumRange(self, left: int, right: int) -> int:
        return self.segement.queryRange(left,right)


# Your NumArray object will be instantiated and called as such:
# obj = NumArray(nums)
# obj.update(index,val)
# param_2 = obj.sumRange(left,right)

File: 307. Range Sum Query - Mutable/307. Range Sum Query - Mutable.cc
class NumArray
{
public:
    vector<int> tree;
    vector<int> values;

    NumArray(vector<int> &nums) : tree(tree)
    {
        values = nums;
        tree.push_back(0);
        tree.insert(tree.end(), nums.begin(), nums.end());
        for (int i = 1; i <= tree.size(); i++)
        {
            int p = i + (i & -i);
            if (p < tree.size())
            {
                tree[p] = tree[p] + tree[i];
            }
        }
    }
    void update(int index, int val)
    {
        int aux = val - values[index];
        values[index] = val;
        index++;
        while (index < tree.size())
        {
            tree[index] += aux;
            index += index & -index;
        }
    }

    int sum(int i)
    {
        int sum = 0;
        while (i > 0)
        {
            sum += tree[i];
            i -= i & -i;
        }
        return sum;
    }

    int sumRange(int left, int right)
    {
        return sum(right + 1) - sum(left);
    }
};

/**
 * Your NumArray object will be instantiated and called as such:
 * NumArray* obj = new NumArray(nums);
 * obj->update(index,val);
 * int param_2 = obj->sumRange(left,right);
 */